### PR TITLE
SAS-3: WIP

### PR DESCRIPTION
**Pull request by darshilshah2017**
WIP

## Summary by CodeRabbit

- New Features
  - Enable PostgreSQL-backed data storage with Liquibase-managed migrations.
  - Introduce partitioned storage for daily equity data to improve performance and scalability.

- Chores
  - Add necessary runtime dependencies and application configuration for database connectivity and migrations.

- Tests
  - Integrate Testcontainers-based PostgreSQL for integration testing.
  - Add tests covering data persistence, retrieval, and duplicate-key constraint behavior.

### CODE REVIEW

File: application/src/main/java/com/github/screener/entity/DailyEquityData.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+    @Column(name = "ded_last_price", nullable = false)
+    private BigDecimal lastPrice;
+
+    @Column(name = "ded_close_price", nullable = false)
+    private BigDecimal closePrice;
+
+    @Column(name = "ded_average_price", nullable = false)
+    private BigDecimal averagePrice;
+
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/main/java/com/github/screener/entity/DailyEquityData.java:36]: The precision and scale are already defined in the SQL file containing the CREATE TABLE scripts. Do we still need to specify them again in the entity class annotations?

File: application/src/test/java/com/github/screener/integration/testcontainers/postgres/PostgresTestContainerContextCustomizerFactory.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PostgresTestContainerContextCustomizerFactory implements ContextCustomizerFactory {
+    private static final String POSTGRES_DEFAULT_IMAGE = "huntress/postgres-partman";
+
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/test/java/com/github/screener/integration/testcontainers/postgres/PostgresTestContainerContextCustomizerFactory.java:25]: I can include this change in a future release if and when the need arises. For now, I'm resolving the issue without it

File: application/src/test/resources/META-INF/spring.factories [hunk 3]
```diff
@@ -0,0 +1,2 @@
+org.springframework.test.context.ContextCustomizerFactory=\
+com.github.screener.integration.testcontainers.postgres.PostgresTestContainerContextCustomizerFactory
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/test/resources/META-INF/spring.factories:2]: I can include this change in a future release if and when the need arises. For now, I'm resolving the issue without it

File: application/etc/database/SAS-3B.sql [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    ded_previous_close NUMERIC(12,2) NOT NULL,
+    ded_open_price NUMERIC(12,2) NOT NULL,
+    ded_high_price NUMERIC(12,2) NOT NULL,
+    ded_low_price NUMERIC(12,2) NOT NULL,
+    ded_last_price NUMERIC(12,2) NOT NULL,
+    ded_close_price NUMERIC(12,2) NOT NULL,
+    ded_average_price NUMERIC(12,2) NOT NULL,
+    ded_total_traded_quantity INTEGER NOT NULL,
+    ded_turnover INTEGER NOT NULL,
+    ded_number_of_trades INTEGER NOT NULL,
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/etc/database/SAS-3B.sql]: Updated to BIGINT

File: application/src/test/java/com/github/screener/repository/DailyEquityDataRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                4500, 60000, 60.00);
+
+        repository.saveAll(List.of(tcsDataToday, tcsDataTomorrow));
+
+        List<DailyEquityData> dailyEquityDataList = repository.findAllById(List.of(
+                new DailyEquityData.DailyEquityDataId(SYMBOL, today),
+                new DailyEquityData.DailyEquityDataId(SYMBOL, today.plusDays(1))
+        ));
+        assertThat(dailyEquityDataList).hasSize(2);
+    }
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/test/java/com/github/screener/repository/DailyEquityDataRepositoryTest.java:57]: But aren't those values already specified in the findAllById method call?

File: application/src/test/java/com/github/screener/StockAnalysisApplicationTest.java [hunk 6]
```diff
@@ -1,13 +1,17 @@
...
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.context.ApplicationContext;
+import org.springframework.test.context.ActiveProfiles;
 
 import static org.junit.jupiter.api.Assertions.*;
 
 @SpringBootTest
+@PostgresTestContainer
+@ActiveProfiles("test")
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/test/java/com/github/screener/StockAnalysisApplicationTest.java:14]: I can include this change in a future release if and when the need arises. For now, I'm resolving the issue without it

File: application/src/main/resources/application.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+spring:
+  datasource:
+    url: jdbc:postgresql://localhost:5432/stock_analysis?currentSchema=stock_analysis
+    username: stockanalysisservice
+    password: ywY3a2to
```
[the remark this answers is not in the document]
Comment (darshilshah2017) [reply application/src/main/resources/application.yml:5]: I feel it's okay to expose the DB password in the local environment, but for dev, QAT, and prod, we'll store the passwords in secrets
